[WIP] GUI: SettingsDialog refactoring and per game config (ignore this for now)

GhettoGirl opened the pull request:
**Work in progress, creating a pull request just for discussions.**

--

Tasks ```SettingsDialog```:
- [x] Rewrite settings dialog, fix UI bugs/glitches, enable real resize capability.
  - [x] Get rid of the *D3D Adapter* setting in Linux and Mac OS X using macro magic.
  - [x] Show *D3D Adapter* only when *D3D12* is selected.
  - [x] Set a minimal size policy to prevent cut offs; overall sizing improvements.
- [x] Implement modes: ```Global, Game``` to show only a subset of options for the game configuration.
- [ ] Implement reading and writing of settings.
- [ ] Reimplement library search feature.
- [ ] Implement Vulkan GPU selector. [needs update of the YAML config and the Vulkan backend aswell]
  - [x] Show Vulkan GPU selector only when Vulkan is selected.
  - [x] Vulkan GPU selector: GUI config
- [x] Clean up the auto generated code from wxFormBuilder and make it readable.
- [ ] *Optimal:* Find and fix bugs and optimize.

--

Let my know if you have any suggestions or feedback :)

Please let me know how the dialog looks in windows. Use **[this screenshots](https://mega.nz/#F!m0BzVLrY!mcMOJqrx9srtWEfnIolU_w)** for comparison. Thanks you :)

raven02 commented:
One suggestion , only display D3D adapter when D3D12 is selected .

GhettoGirl commented:
@raven02 thanks, will add this to the task list :)

paulsapps commented:
It would be nice if the modules selection part used up all of the dialog space, seems to be lots of unused space in that tab

GhettoGirl commented:
@paulsapps you mean when the dialog is resized? , because in the initial size everything looks good to me. i'm still investigating why wxWidgets acts so dumb, even using a form builder now I still don't understand what could be wrong.

paulsapps commented:
Latest version looks better than whatever I was using actually, but I mean between the module search text bottom and the bottom of the dialog there is a big gap? So the actual list could probably use this space?

kd-11 commented:
We actually need to add a GPU selector for vulkan as well instead of just using GPU 0. I wasnnt able to finish up that little bit, but it could help users with multi-GPU or optimus setups.

GhettoGirl commented:
@kd-11 Thats a great idea, added this to the task list. This would also require to update the YAML config and the Vulkan backend aswell.

raven02 commented:
Meanwhile , i cannot see the two following options used ever .

3D Monitor
Log shader programs (i think we always log shader for all backends)

kd-11 commented:
Well, in future, regular users may not require logging of shaders although that is still a long way off.

Nekotekina commented:
Any progress with this?

GhettoGirl commented:
I've a lot of trouble with the current YAML configuration, whatever I'm trying todo the per game YAML always has parts from the global config in it.

I just don't get it how I can remove parts from the config, or even better read just the indented values.

And yes I read the yaml-cpp documentation and tried a lot of things.

GhettoGirl commented:
Currently I'm just loading the whole configuration, even for the per game config which isn't a good idea, on the other side I don't know how I can just read the values I need. What I'm trying to archive is to read default values from the global user configuration and than override them with the available config for the game (if any) and load this in the GUI - and than store everything back.

The dialog itself looks good otherwise and I also removed the *3D Monitor* option since it got removed in this commit 786664f94d4545e16a4f1c09e5b29a702c53f61e

GhettoGirl commented:
I rebased and refactored the dialog a bit for now, there where merge conflicts because of the commit mentioned.

To be honest, I need some advice to understand the YAML configuration.

GhettoGirl commented:
I finished the task **[Implement modes: Global, Game to show only a subset of options for the game configuration.]** just now. Which means I can start to implement the configuration loading and saving now.

The code is simplified with the help of some macros. I reordered and cleaned up the generated code from wxFormBuilder. The code is very short and simple now. Give some feedback if you're all ok with the changes :)

The rest will be implemented using events: ```Bind(...)```

GhettoGirl commented:
Implemented GPU selector for D3D and Vulkan.
 - D3D is done and should work. (untested, I'm not sure if this code whould actually work in a virtual machine)
 - To make Vulkan also work, available GPUs only needs to inserted into a list. I've no idea of graphic development, so I can't add this by myself.

Nekotekina commented:
I see you are trying to use predefined set of options for "Game" mode... Doing this shouldn't be hard and something similar actually existed some time ago, but my "Ideal" vision is an ability to manually specify which options you want to save. For checkbox it could be solved using three states (on, off, gray state to not save). For comboboxe, by additional selection item. For others, I'm not sure how to specify it, maybe additional checkboxes. But it's just an opinion.

Alternatively, some "clever" config tree comparison could be implemented to save only changed options. May be unreliable.

Binding events may not be the best solution. In current implementation, values are read only after the dialog is closed by pressing "OK".

About YAML, the example of writing full tree is found in Utilities/Config.cpp, storing only parts of it may be tricky. I'll look at it again later.

vlj commented:
@GhettoGirl IMHO backend should exposes functions to enumerate GPU. 
For D3D12 this is currently done somewhere in the gui code but it's suboptimal.

Enumerating GPU requires that backend is loaded and enabled. For OpenGL it's always the case (except maybe on Linux if mesa is not installed) ; for D3D12 one should load the dll first to ensure they're available (rpcs3 can be run on windows 7) then GPU should be enumerated (there is always at least one "gpu" present, "basic renderer" which is a software fallback); Vulkan is statically linked into rpcs3 so it's always loaded but it needs working drivers which can be missing (for instance no supported gpu).

I think GSRender interface should be extended with a 2 step initialisation procedures, one that loads the dll and initialize "instance" and which can fails, and another one that create everything needed for rendering.

GhettoGirl commented:
For some reason appveyor fails to clone into the ffmpeg submodule :confused: 
I guess I messed something up with my branch, because other PRs work fine.

Nekotekina commented:
@vlj I don't see any reason to extend GSRender for this purpose. It only complicates things.

Nekotekina commented:
@GhettoGirl I restarted appveyor, it has another error.

Nekotekina commented:
Just remembered: wxT is compatibility thing, we don't need to use it.

raven02 commented:
It looks strange on windows platform though it is WIP. And , cannot save the configuration yet.

![1](https://cloud.githubusercontent.com/assets/3000282/17316343/af4ed866-58a5-11e6-8196-2f1abaceed73.jpg)
![2](https://cloud.githubusercontent.com/assets/3000282/17316344/af565550-58a5-11e6-98c8-05525dd7233d.jpg)

GhettoGirl commented:
@raven02 I will change the design of the dialog anyway. And there is no config reading/saving implemented yet, just FYI.

paulsapps commented:
Apart from the space between the frame and the buttons what part is strange?

GhettoGirl commented:
I'm closing this because I'm done with wxWidgets and its shitty layout design. I can't effort working on this dialog any longer :disappointed: Besides I'm active on other projects right now.

Sorry about that

danilaml commented:
@GhettoGirl such a shame. Didn't need to be perfect, just better than it is currently would be enough.
